general/array.hpp

```diff
@@ -103,14 +98,17 @@ class Array
    explicit inline Array(std::initializer_list<CT> values);
 
    /// Move constructor ("steals" data from 'src')
-   inline Array(Array<T> &&src) : Array() { Swap(src, *this); }
+   Array(Array<T> &&src) = default;
 
    /// Destructor
    inline ~Array() { data.Delete(); }
 
-   /// Assignment operator: deep copy from 'src'.
+   /// Copy assignment operator: deep copy from 'src'.
    Array<T> &operator=(const Array<T> &src) { src.Copy(*this); return *this; }
 
+   /// Move assignment operator
+   Array<T> &operator=(Array<T> &&src) = default;
```

Comment:
This is incorrect -- the current content of the `Array` is leaked.

Reply:
For the move assignment, I think we should keep the `src.DeleteAll()` and we just need to add a check for self-assignment:
```c++
template <class T>
inline Array<T> &Array<T>::operator=(Array<T> &&src)
{
   if (this == &src) { return *this; }
   std::swap(data, src.data);
   std::swap(size, src.size);
   src.DeleteAll();
   return *this;
}
```

For the move construction: if we use the move assignment, we first need to make sure `*this` is initialized. Instead, a simpler implementation is this:
```c++
template <typename T>
inline Array<T>::Array(Array<T> &&src)
   : data(std::move(src.data)), size(src.size)
{
   src.size = 0;
}
```

Reply:
Done in 5a7edc9548de0d9aaa82be63e92533cb19394534.